Warn user when unsound flags are enabled.

NlightNFotis opened the pull request:
Hello,

This is a draft PR that enables warning a user when `--reachability-slice` and `--full-slice`
are used in `cbmc` or `goto-instrument`.

It also adds a small doc for explaining the rationale for this, and how the warnings look.

For more discussion on this, please have a look at #6480

NlightNFotis committed:
Issue warnings when options may yield unsound verification results

Also warn about options that aren't unsound by design, but are
experimental in nature.

NlightNFotis committed:
Explain unsound and experimental options in documentation.

Some use of CBMC or JBMC can yield verification results that must not be
taken at face value.

Co-authored-by: Michael Tautschnig <tautschn@amazon.com>

tautschnig commented on src/cbmc/cbmc_parse_options.cpp:
```diff
@@ -952,6 +952,8 @@ bool cbmc_parse_optionst::process_goto_program(
   if(options.get_bool_option("reachability-slice"))
   {
+    log.warning() << "WARNING: Unsound option --reachability-slice"
+                  << messaget::eom;
```
This option isn't fundamentally unsound, it's just that #6505 needs to be merged to fix a bug.

tautschnig replied:
The above PR has been merged, so I don't think this applies any longer.

tautschnig reviewed (changes requested):
Thank you for kicking off work to annotate options and document what's going on. For "reachability-slice," however, I'd rather see us merge the fix.

TGWDB reviewed (changes requested):
As per the discussion we had in #6480 I thought the conclusion was to be explicit in how an option can be "unsound" as either *Spurious counter example to an assertion* or *Wrong proof to an assertion*. Can we update the messages and documentation here to say what can go wrong.

NlightNFotis commented:
Hi @tautschnig, some updates with regard to this:

1) I have removed the `--reachability-slice` mark as requested.
2) I have explored the behaviour here a bit, and can report the following:

If we enable any of the slice flags on a global level for a regressions test-suite run, for example `--full-slice` for all tests under `regression/cbmc/` we do get a number of failures that appear to be unexpected (to me at least). 

The failures appear to be flipping of verification results (from success to failure), and tend to refer to values not present in the trace. I have noticed that what happens in these cases is that we are asserting against values that appear to be assigned the result of a function call, but the slice options appear to eliminate the function body.

This is what happens in the case of a `--full-slice`, and which I've done some digging into. Test failures for similar reasons appear to be the result of other flags activated on a global basis, (for instance, `--reachability-slice` or `--slice-formula`) but I'm not sure if it's the exact same behaviour driving this.

I have done a number of quick fix-like experiments, as well as a few other structural changes to the full-slice but couldn't get anything to work properly (by this I mean passing all the tests), which signals a lack of understanding of the slicing internals from my end. But this appears to be genuine wrong behaviour (I expect any slicer to effectively preserve at least the direct path to an assignment).

NlightNFotis commented:
I have left a more detailed comment outlining the current issue in #260 .

NlightNFotis commented:
Hi @tautschnig, I have changed this to now also mark `--reachability-slice` as unsound, based on comments made during one of our regular catchup meetings.

Can you please confirm this is the case?

Could you also have another look at this at your earliest convenience, given that this is blocked by you?

NlightNFotis commented:
Hi @tautschnig, ping here - is this possible to review this so that we can get it merged?

tautschnig commented:
> Hi @tautschnig, I have changed this to now also mark `--reachability-slice` as unsound, based on comments made during one of our regular catchup meetings.
> 
> Can you please confirm this is the case?
> 
> Could you also have another look at this at your earliest convenience, given that this is blocked by you?

With apologies for the late response: is it still the case that `--reachability-slice` is knowingly unsound/producing wrong results? I am aware that `--full-slice` has issues and marking that one "unsound" appears fine.

tautschnig commented on doc/cprover-manual/unsound_options.md:
```diff
@@ -0,0 +1,37 @@
+  In particular, `--full-slice` appears to lead to spurious counter examples,
+  because values that get assigned by a function whose body gets sliced out are
+  no longer present in the trace, but still result in flipped verification results.
+* `--reachability-slice` has been reported unsound at least once, in issue [cbmc#6394](https://github.com/diffblue/cbmc/issues/6394).
+  The state of this option these days is that this flag *should not* contain significant
+  soundness bugs, but it has a number of potential issues that pop up everyone now and then,
+  which renders it unsuitable for production use.
```
I think we might want to list other options in here, but I'm not sure `--reachability-slice` still warrants this warning. On the other hand, `--aggressive-slice` would fit in here.

tautschnig commented on doc/cprover-manual/unsound_options.md:
```diff
@@ -0,0 +1,37 @@
+# Unsound CLI options for various tools
+
+Be advised that the following command line options to `cbmc` and `goto-instrument`
+have been reported to be unsound:
```
Can I ask that we start with a bit more context (reflecting the fact that https://github.com/diffblue/cbmc/issues/6480 had some extensive discussion)? I don't think there even is a universally agreed definition of "unsound," so please try to start with a summary of the conclusion of #6480.

NlightNFotis replied:
Added a paragraph with the conclusions from #6480

tautschnig commented on doc/cprover-manual/unsound_options.md:
````diff
@@ -0,0 +1,37 @@
+[...]
+```
+
+Be advised some of these options may be activated by auxiliary options, for instance,
+`--aggressive-slice` seems to lead to a path that performs the transformations that
+`--reachability-slice` does in `goto-instrument`.
````
We should not do this via documentation, but instead need to make sure that the code always issues a warning when one of the unsound code paths is entered.

tautschnig commented on src/goto-instrument/goto_instrument_parse_options.cpp:
```diff
@@ -1806,6 +1814,7 @@ void goto_instrument_parse_optionst::instrument_goto_program()
     goto_model.goto_functions.update();
 
     log.status() << "Performing a reachability slice" << messaget::eom;
+
```
Whitespace change?

tautschnig commented on src/goto-instrument/goto_instrument_parse_options.cpp:
```diff
@@ -1743,7 +1749,9 @@ void goto_instrument_parse_optionst::instrument_goto_program()
 
+    log.warning() << "WARNING: Unsound option --full-slice" << messaget::eom;
     log.status() << "Performing a full slice" << messaget::eom;
+
```
Whitespace change?

NlightNFotis commented:
Hi @tautschnig,

I believe I have addressed all of the comments.

I have also removed the explanation for the `--reachability-slice` from the documentation, along with the `--aggressive-slice` option.

I'm not aware of any issues with the `--aggressive-slice` (the references before was due to a transitive relation to `--reachability-slice`) and I removed the references to it. If you are aware of any issues with it let me know and I can add it back in with a more elaborate explanation.

jimgrundy commented on doc/cprover-manual/unsound_options.md:
```diff
@@ -0,0 +1,52 @@
+We concluded that *unsound* in this context is a proxy for the following two behaviours
+that we want to avoid:
+
+* **A spurious counter example to an assertion**, which gives a proof that the assertion
```
I think a lot of people would fail to recognize this as a soundness problem. But if you cast it another way - that the tool may report a coverage property (or line of code) as reachable when, in fact, it is not - then I think more people will see that this is just as serious as a spurious proof. I would prefer we state it that way.

NlightNFotis commented:
Hi,

@jimgrundy @tautschnig 

The comments you have left have now been addressed. Could you please have another look at your earliest convenience? Cheers.

tautschnig reviewed (approved):
We can merge this as is, but I am wondering whether we might want to speak of "Unexpectedly" unsound options? Examples of transformations that incur unsoundness are `--partial-loops` (may yield spurious counterexamples) or `--aggressive-slicer` (may yield both spurious counterexamples as well as wrong proof). Both of these, AFAIK, work exactly as their documentation suggests, but do yield those unsound results.

jimgrundy reviewed (changes requested):
Hang on... there are options beyond the ones documented here that are known to cause unsoundness? Surely those must be documented here. It's almost worse than the present situation to go ahead as is, because if I now believe the tool will warn me when I ask for something unsound and I don't get a warning then I'll be even more convinced of the validity of the results I get.

tautschnig commented:
> Hang on... there are options beyond the ones documented here that are known to cause unsoundness? Surely those must be documented here. It's almost worse than the present situation to go ahead as is, because if I now believe the tool will warn me when I ask for something unsound and I don't get a warning then I'll be even more convinced of the validity of the results I get.

@jimgrundy I have now added a further commit as discussed out of band.

jimgrundy reviewed (changes requested):
I like these changes even more, but, since I'm asking
- It feels now like there should be a heading and _short_ section on options that are unsound by design (which probably isn't the most catchy name for them). If we could give an example or two of these + state that these options will give a warning when used.
Do we know how many such options there are? If we don't know or there are lots then we could defer adding the warning until we tackle soundness more completely, but if there if there are just a couple could we sneak the warnings in now?

tautschnig committed:
Synchronise help output with actual support for --fp-reachability-slice

Neither CBMC nor JBMC actually interpret this option, despite it being
advertised in help output (and man pages). Conversely, goto-instrument
has support for `--reachability-slice-fb`, but it wasn't available on
the command line.

tautschnig committed:
Add missing newline in help output

We only ended up with reasonable output as the next line was a newline.

tautschnig committed:
Remove (broken) code contracts link from CPROVER manual

The underlying files were moved in e22a788588569b7a52f5.

tautschnig committed:
CPROVER manual: consistent formatting of headings

Use heading levels that match the nesting level within the overall
index.

tautschnig commented:
> I like these changes even more, but, since I'm asking
> 
>     * It feels now like there should be a heading and _short_ section on options that are unsound by design (which probably isn't the most catchy name for them). If we could give an example or two of these + state that these options will give a warning when used.

Done.

>       Do we know how many such options there are? If we don't know or there are lots then we could defer adding the warning until we tackle soundness more completely, but if there if there are just a couple could we sneak the warnings in now?

I believe I have covered all of them for CBMC. JBMC has a more elaborate environment generation facility, which might have its own share of risks. Perhaps @peterschrammel can comment on what I've covered so far?

peterschrammel commented on doc/cprover-manual/unsound_options.md:
```diff
@@ -0,0 +1,67 @@
+We concluded that *unsound* in this context is a proxy for the following two behaviors
+that we want to avoid:
+
+* **A spurious counter example to an assertion**, which means that the tool may report
```
counterexample
